### PR TITLE
introduction to rest api documentation

### PR DESCRIPTION
**Pull request by vicancy**
@chenkennt @superyyrrzz @qinezh @vwxyzh

### CODE REVIEW

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+DocFX now supports generating documentation from REST APIs following [Swagger specification](http://swagger.io/specification/) version 2.0.
+
+The Swagger RESTful API files *MUST* be end with `.swagger.json` or `.swagger2.json` or `_swagger.json` or `_swagger2.json` so that these files are recognized as REST API files in DocFX.
+
+One Swagger API file generates one HTML file. For example. a file `contacts.swagger.json` generates file naming `contacts.swagger.html`.
+
+Basic structure
+--------------
+A single Swagger API file is considered as an unique **Entity** containing multiple **API**s. The **UID**(Unique IDentifier) for this **Entity** is defined as the combination of `host`, `basePath` and `info.title` with `/` as seperator. For example, the following Swagger API file has **UID** equals to `microsoft.com/docfx/Contacts`:
```
Comment (superyyrrzz) [inline Documentation/tutorial/intro_rest_api_documentation.md]: an unique => a unique ?

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+> It is recommended that user provides a well-formed `operationId` name. We suggest that the `operationId` is one word in camel case or snake case.
+
+HTML layout
+--------------
+By default, the generated HTML file lists all the **API**s inside the **Entity** in the order defined in the Swagger REST file.
+
+You can use *Overwrite File*s to redefine the layout of the **API**s and adding more information to the **Entity** and **API**.
+
+### *Overwrite File*s
+*Overwrite File*s are markdown files with multiple *Overwrite Section*s starting with YAML header block. A valid YAML header for an *Overwrite Section* *MUST* take the form of valid [YAML](http://www.yaml.org/spec/1.2/spec.html) set between triple-dashed lines and start with property `uid`. Here is a baisc example for an *Overwrite Section*:
```
Comment (superyyrrzz) [inline Documentation/tutorial/intro_rest_api_documentation.md]: baisc => basic

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+```md
+---
+uid: microsoft.com/docfx/Contacts
+footer: *content
+---
+Footer for `microsoft.com/docfx/Contacts`
+```
+
+If the value is set to `*content`, it means that the value is the markdown content following YAML header. If footer is defined, the content will be append to the last section of the **Entity** or **API**. It is usually used to define **See Also** or **Additional Resources** for the documentation.
```
Comment (superyyrrzz) [inline Documentation/tutorial/intro_rest_api_documentation.md]: append => appended

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
+Introduction to REST API documentation
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: documentation -> Documentation

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+===============================
+
+Introduction
+-------------
+
+DocFX now supports generating documentation from REST APIs following [Swagger specification](http://swagger.io/specification/) version 2.0.
+
+The Swagger RESTful API files *MUST* be end with `.swagger.json` or `.swagger2.json` or `_swagger.json` or `_swagger2.json` so that these files are recognized as REST API files in DocFX.
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: also add "." for _swagger,json and _swagger.2.json

Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: ~~be~~ end with

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+One Swagger API file generates one HTML file. For example. a file `contacts.swagger.json` generates file naming `contacts.swagger.html`.
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: should be contracts.html?

File: Documentation/tutorial/intro_rest_api_documentation.md (hunk 1 above)
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: use our term in yaml spec? entity -> item?

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+}
+```
+
+> It is recommended that user provides a well-formed `operationId` name. We suggest that the `operationId` is one word in camel case or snake case.
+
+HTML layout
+--------------
+By default, the generated HTML file lists all the **API**s inside the **Entity** in the order defined in the Swagger REST file.
+
+You can use *Overwrite File*s to redefine the layout of the **API**s and adding more information to the **Entity** and **API**.
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: adding -> add

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+---
+uid: microsoft.com/docfx/Contacts
+sections:
+  - microsoft.com/docfx/Contacts/get_contacts
+  - microsoft.com/docfx/Contacts/add_contacts
+---
+Further description for `microsoft.com/docfx/Contacts`
+```
+
+### Add footer
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: This is a general functionality, also add it to our DFM spec.

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+You can also define the footer for an **Entity** or **API** using the following syntax:
+
+```md
+---
+uid: microsoft.com/docfx/Contacts
+footer: *content
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: Add a link to YAML anchor spec to explain why we choose to use *

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+By default, the generated HTML file lists all the **API**s inside the **Entity** in the order defined in the Swagger REST file.
+
+You can use *Overwrite File*s to redefine the layout of the **API**s and adding more information to the **Entity** and **API**.
+
+### *Overwrite File*s
+*Overwrite File*s are markdown files with multiple *Overwrite Section*s starting with YAML header block. A valid YAML header for an *Overwrite Section* *MUST* take the form of valid [YAML](http://www.yaml.org/spec/1.2/spec.html) set between triple-dashed lines and start with property `uid`. Here is a baisc example for an *Overwrite Section*:
+
+```md
+---
+uid: microsoft.com/docfx/Contacts
```
Comment (hellosnow) [inline Documentation/tutorial/intro_rest_api_documentation.md]: also introduce "tryFeature" in yaml header?

File: Documentation/tutorial/intro_rest_api_documentation.md (hunk 5 above)
Comment (vicancy) [inline Documentation/tutorial/intro_rest_api_documentation.md]: "." is not required for `_swagger2.json` and `_swagger.json`. Actually `contacts_swagger2.json` is aad's naming convention.

File: Documentation/tutorial/intro_rest_api_documentation.md (hunk 10 above)
Comment (vicancy) [inline Documentation/tutorial/intro_rest_api_documentation.md]: We don't have tryFeature in default template though...

File: Documentation/tutorial/intro_rest_api_documentation.md (hunk 6 above)
Comment (vicancy) [inline Documentation/tutorial/intro_rest_api_documentation.md]: makes sense. Requires code change.

File: Documentation/tutorial/intro_rest_api_documentation.md (hunk 8 above)
Comment (vicancy) [inline Documentation/tutorial/intro_rest_api_documentation.md]: Will provide a doc for overwrite fils

**Comment by vicancy** [comment]
Updated, please review @chenkennt

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        "responses": {
+        },
+        "operationId": "get_contacts"
+      }
+    },
+  }
+}
+```
+
+> It is recommended that user provides a well-formed `operationId` name. We suggest that the `operationId` is one word in camel case or snake case.
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: camelCase or snake_case

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+---
+uid: microsoft.com/docfx/Contacts
+---
+Further description for `microsoft.com/docfx/Contacts`
+```
+
+The `uid` value *MUST* match the `uid` of the **API** or **Item** that you want to overwrite. The content following YAML header is the additional markdown description for the **API** or **Item**. By default, it is transformed to HTML and appended below the description of the **Item** or **API**.
+
+### Redefine the layout
+You can redefine the **API**s layout in the HTML page using `sections` property. For example, the following *Overwrite Section* specifies that these two **API**s and only these two **API**s with `uid` equals to `get_contacts` and `add_contacts` are shown in the HTML page in order:
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: "only APIs" should be enough

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+If `footer` is set, the content from `footer` will be appended to the last section of the **Item** or **API**. It is usually used to define **See Also** or **Additional Resources** for the documentation.
+
+### Add other metadata
+You can define your own metadata with YAML header. This functionality is quite useful when your own template is used.
+
+When the key of the metadata is already preserved by DocFX, for example, `summary`, the value of `summary` will be overwritten. You can also overwrite complex types, for example, `description` of a `parameter`. Make sure the data structure of the provided metadata is consistent with the one defined in DocFX, otherwise, DocFX is unable to cast the value and throws exception.
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: fail. don't use "throw exception" in user doc..

**Comment by chenkennt** [comment]
:-1:

**Comment by vicancy** [comment]
@chenkennt Updated.

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  "host": "microsoft.com",
+  "basePath": "/docfx",
+  "schemes": [
+    "https"
+  ]
+}
+```
+
+The Swagger API **Item** contains multiple **API**s as its children. An **API** is an [Operation Object](http://swagger.io/specification/#operationObject) defined in [Path Item Object](http://swagger.io/specification/#pathItemObject). The **UID**(Unique IDentifier) for this **API** is defined as the combination of the **UID** of the **Item** and the `operationId` of the [Operation Object](http://swagger.io/specification/#operationObject). For example, the following `get_contacts` operation has **UID** equal to `microsoft.com/docfx/Contacts/get_contacts`:
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: item sounds weird here, use Swagger API file?

**Comment by vicancy** [comment]
@chenkennt  updated.
@hellosnow Version should also be considered as part of UID, I will update code accordingly, meanwhile, we need to update REST content...

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+DocFX now supports generating documentation from REST APIs following [Swagger specification](http://swagger.io/specification/) version 2.0.
+
+The Swagger RESTful API files *MUST* end with `.swagger.json` or `.swagger2.json` or `_swagger.json` or `_swagger2.json` so that these files are recognized as REST API files in DocFX.
+
+One Swagger API file generates one HTML file. For example. a file `contacts.swagger.json` generates file naming `contacts.swagger.html`.
+
+Basic structure
+--------------
+A single Swagger API file is considered as a unique **File** containing multiple **API**s. The **UID**(Unique IDentifier) for the **Item** is defined as the combination of `host`, `basePath`, `info.title` and `info.version` with `/` as seperator. For example, the following Swagger API file has **UID** equals to `microsoft.com/docfx/Contacts/1.6`:
```
Comment (chenkennt) [inline Documentation/tutorial/intro_rest_api_documentation.md]: still one Item -> File

File: Documentation/tutorial/intro_rest_api_documentation.md [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  "host": "microsoft.com",
+  "basePath": "/docfx",
+  "schemes": [
+    "https"
+  ]
+}
+```
+
+A Swagger API **File** contains multiple **API**s as its children. An **API** is an [Operation Object](http://swagger.io/specification/#operationObject) defined in [Path Item Object](http://swagger.io/specification/#pathItemObject). The **UID**(Unique IDentifier) for this **API** is defined as the combination of the **UID** of the **File** and the `operationId` of the [Operation Object](http://swagger.io/specification/#operationObject). For example, the following `get_contacts` operation has **UID** equal to `microsoft.com/docfx/Contacts/1.6/get_contacts`:
```
Comment (hellosnow) [inline Documentation/tutorial/intro_rest_api_documentation.md]: have a question, when to use **File**? Some is "Swagger API file" and some "Swagger API **File**"?

Comment (vicancy) [inline Documentation/tutorial/intro_rest_api_documentation.md]: Nice catch, use `REST API **File**` instead

**Comment by vicancy** [comment]
@chenkennt Updated to include code changes, please review.
@hellosnow Hi Jessie, the commit contains 2 changes, both have influence to REST template:
1. `version` is added to `uid` => REST content need update
2. `contacts_swagger2.json` should be transformed to `contacts.html`. => the template to save the original swagger file should be renamed to `RestApi.swagger.json`.

**Comment by chenkennt** [comment]
:+1:

**Pull request merged**